metsim/metsim.py

```diff
@@ -42,71 +47,143 @@ def __init__(self, params:dict):
-        domain.close()
-
-        # If we are outputting netcdf, intiialize the file
-        if MetSim.params.get('out_format', 'netcdf') == 'netcdf':
-            self.init_netcdf()
-
-
-    def run(self, forcings):
+        self.domain = Dataset(MetSim.params['domain'], 'r')
+        self.domain_lats = np.array(self.domain['lat'])
+        self.domain_lons = np.array(self.domain['lon'])
+        self.elev = np.array(self.domain['elev'])
+
+        # Create the data structures for writeout
+        self.manager = Manager()
+        self.queue = self.manager.Queue()
+
+
+    def launch(self, job_list):
+        """Farm out the jobs to separate processes"""
+        # Used to dispatch to the correct preprocessing functions
+        in_preprocess = {"ascii" : self.ascii_in_preprocess,
+                         "binary" : self.binary_in_preprocess,
+                         "netcdf" : self.netdf_in_preprocess}
+        met_data = in_preprocess[MetSim.params['in_format']](job_list)
+
+        # Start up the IO Process
+        io_process = Process(target=self.create_io_process, args=(self.queue,))
+        io_process.start()
+
+        # Split the jobs up into groups based on the number
+        # of desired processes
+        self.locations = np.array_split(np.array(self.locations),  MetSim.params['nprocs'])
+        process_handles = [Process(target=self.run, args=(met_data, locs, self.queue))
+                           for locs in self.locations]
+
+        # Runs everything
+        for p in process_handles:
+            p.start()
+        for p in process_handles:
+            p.join()
+
+        # Close everything out
+        self.queue.put("done")
+        io_process.join()
+
+
+    def run(self, met_data, locations, queue):
         """
         Kicks off the disaggregation and queues up data for IO
         """
         # Where we will store the results for writing out
         out_dict = {}
         method = MetSim.methods[MetSim.params.get('method', 'mtclim')]
 
-        # Coerce the forcings into a list if it's not one
-        if type(forcings) is not list:
-            forcings = [forcings]
-
-        # Do the forcing generation and dissaggregation if required
-        for forcing in forcings:
-            met_data = self.read(forcing)
-            for i in range(len(met_data.lat)):
-                out_dict[i] = {}
-                for j in range(len(met_data.lon)):
-                    out_dict[i][j] = method.run(met_data.isel(lat=[i], lon=[j])
-                                                .to_dataframe(), MetSim.params,
-                                                disagg=True)
-            # Write out
-            self.write(out_dict)
-
-        if MetSim.params.get('out_format', 'netcdf') == 'netcdf':
-            self.output.close()
+        # Do the forcing generation and disaggregation if required
+        for i, j in locations:
+            print("Processing {} {}".format(i, j))
```

Comment:
OK - but use the logging module to implement verbose mode. It's pretty easy to use and will allow you to configure different levels of verbosity.